src/symbolicalgebra/abstractalgebra.jl

```diff
@@ -0,0 +1,223 @@
+"""
+    SymbolicAlgebra
+
+Abstract supertype for working with symbolic (operator) algebras.
+"""
+abstract type SymbolicAlgebra{T<:Number} end
+
+# Building blocks
+# ---------------
+struct Op{T<:Number,A}
+  id::A
+end
+
+struct Scaled{O,T<:Number}
+  op::O
+  scalar::T
+end
+
+struct Sum{O,T<:Number}
+  terms::Dictionary{O,T}
+  Sum{O,T}() where {O,T} = new{O,T}(Dictionary{O,T}())
+  Sum{O,T}(terms::Dictionary{O,T}) where {O,T} = new{O,T}(terms)
+end
+
+struct Prod{O}
+  factors::Vector{O}
+  Prod{O}() where {O} = new{O}(O[])
+  Prod{O}(factors::Vector{O}) where {O} = new{O}(factors)
+end
+
+struct Kron{O}
+  factors::Vector{O}
+  Kron{O}() where {O} = new{O}(O[])
+  Kron{O}(factors::Vector{O}) where {O} = new{O}(factors)
+end
+
+struct Fun{O}
+  f::Any
```

Comment:
This is something I'd gone back and forth about, what are the implications of this abstract field?

Reply:
Agreed that specifying it as `::Function` wouldn't make much difference. This seems like a reasonable starting point. An alternative design I considered was having different variants of the sum type/ADT for different functions, though that obviously isn't as general. Maybe some functions warrant that if they have specialized functionality (for example, we might want a specialized variant for `exp` since it has particular algebra rules we want to take advantage of, like Trotterization).